### PR TITLE
standardize coroutine context propagation in the execution

### PR DESCRIPTION
**Pull request by dariuszkuc**
### :pencil: Description

Attempts to standardize how we can propagate the coroutine/reactor context created by the server when processing incoming requests to the data fetchers that would resolve the underlying fields.

`GraphQLContext` map was introduced in `graphql-java` 17 as means to standardize usage of GraphQL context (previously it could be any object). We can store the original GraphQL request context in the map and then use it from within the data fetchers.

This is a breaking change as we are changing signatures of `GraphQLContextFactory` and `FunctionDataFetcher`.

### :link: Related Issues

* https://github.com/ExpediaGroup/graphql-kotlin/pull/1336
* https://github.com/ExpediaGroup/graphql-kotlin/issues/1318
* https://github.com/ExpediaGroup/graphql-kotlin/discussions/1300
* https://github.com/ExpediaGroup/graphql-kotlin/issues/1257

### CODE REVIEW

File: servers/graphql-kotlin-server/src/main/kotlin/com/expediagroup/graphql/server/execution/GraphQLContextFactory.kt [hunk 1]
```diff
@@ -16,23 +16,16 @@
 
 package com.expediagroup.graphql.server.execution
 
-import com.expediagroup.graphql.generator.execution.GraphQLContext
+const val GRAPHQL_COROUTINE_CONTEXT_KEY = "GRAPHQL_COROUTINE_CONTEXT"
```
Comment (smyrick) [inline servers/graphql-kotlin-server/src/main/kotlin/com/expediagroup/graphql/server/execution/GraphQLContextFactory.kt:19]: Is this supposed to be `GRAPHQL_COROUTINE_SCOPE`? Maybe we should just use the value defined in the generator package so we don't have it defined in two locations

File: servers/graphql-kotlin-spring-server/src/main/kotlin/com/expediagroup/graphql/server/spring/subscriptions/ApolloSubscriptionSessionState.kt [hunk 2]
```diff
@@ -31,43 +30,24 @@ internal class ApolloSubscriptionSessionState {
...
-        }
-    }
-
-    /**
-     * Save the context created from the factory and possibly updated in the onConnect hook.
-     * This allows us to include some intial state to be used when handling all the messages.
-     * This will be removed in [terminateSession].
-     */
-    fun saveContextMap(session: WebSocketSession, graphQLContext: Map<*, Any>?) {
+    fun saveContextMap(session: WebSocketSession, graphQLContext: Map<*, Any?>?) {
```
Comment (smyrick) [inline servers/graphql-kotlin-spring-server/src/main/kotlin/com/expediagroup/graphql/server/spring/subscriptions/ApolloSubscriptionSessionState.kt:41]: If we are making the breaking change do we want to call it `saveGraphQLContext` to match conventions?

Comment (dariuszkuc) [reply servers/graphql-kotlin-spring-server/src/main/kotlin/com/expediagroup/graphql/server/spring/subscriptions/ApolloSubscriptionSessionState.kt:41]: Yeah was just trying to get it compile first.

File: servers/graphql-kotlin-server/src/main/kotlin/com/expediagroup/graphql/server/execution/GraphQLContextFactory.kt (hunk 1 above)
Comment (dariuszkuc) [reply servers/graphql-kotlin-server/src/main/kotlin/com/expediagroup/graphql/server/execution/GraphQLContextFactory.kt:19]: Actually I'm thinking of dropping the String key in favor of just using classes as keys.

File: servers/graphql-kotlin-server/src/main/kotlin/com/expediagroup/graphql/server/execution/GraphQLServer.kt [hunk 3]
```diff
@@ -39,18 +45,21 @@ open class GraphQLServer<Request>(
...
         val graphQLRequest = requestParser.parseRequest(request)
 
-        return if (graphQLRequest != null) {
-            val context = contextFactory.generateContext(request)
-            val graphQLContext = contextFactory.generateContextMap(request)
+        if (graphQLRequest != null) {
+            val contextMap = contextFactory.generateContextMap(request) ?: emptyMap<Any, Any?>()
+
+            val customContext: CoroutineContext = contextMap[GRAPHQL_COROUTINE_CONTEXT_KEY] as? CoroutineContext ?: EmptyCoroutineContext
+            val graphQLExecutionScope = CoroutineScope(coroutineContext + customContext + SupervisorJob())
```
Comment (jgorman-exp) [inline servers/graphql-kotlin-server/src/main/kotlin/com/expediagroup/graphql/server/execution/GraphQLServer.kt:55]: Not an expert on coroutines, but does this need to be cancelled at the end of request processing?  Was just reading through this:  https://kotlin.github.io/kotlinx.coroutines/kotlinx-coroutines-core/kotlinx.coroutines/-coroutine-scope/index.html

Comment (dariuszkuc) [reply servers/graphql-kotlin-server/src/main/kotlin/com/expediagroup/graphql/server/execution/GraphQLServer.kt:54]: When we return the value from this method then all children coroutines will already complete so there is nothing to cancel.

**Comment by TiKevin83** [comment]
Ran into #1300 testing this library in a Kotlin project and was able to confirm that this PR resolves the issues we were seeing with the spring security context being empty when coming from the GraphQL endpoints!

**Comment by wallind** [comment]
> Ran into #1300 testing this library in a Kotlin project and was able to confirm that this PR resolves the issues we were seeing with the spring security context being empty when coming from the GraphQL endpoints!

to be even more specific we were able to get things working with manually built jars specifically from this commit https://github.com/ExpediaGroup/graphql-kotlin/pull/1349/commits/87fc89ce7b9abaee2025b273c14a6f80af90c692

**Comment by dariuszkuc** [comment]
Updated to keep the deprecated custom GraphQL context around (will drop it in v7) to provide easier migration path.

File: servers/graphql-kotlin-spring-server/src/main/kotlin/com/expediagroup/graphql/server/spring/execution/SpringGraphQLContextFactory.kt [hunk 4]
```diff
@@ -25,8 +26,13 @@ import org.springframework.web.reactive.function.server.ServerRequest
...
 /**
- * Basic implementation of [SpringGraphQLContextFactory] that returns [SpringGraphQLContext]
+ * Basic implementation of [SpringGraphQLContextFactory] that populates Apollo tracing header.
  */
-class DefaultSpringGraphQLContextFactory : SpringGraphQLContextFactory<SpringGraphQLContext>() {
-    override suspend fun generateContext(request: ServerRequest) = SpringGraphQLContext(request)
+open class DefaultSpringGraphQLContextFactory : SpringGraphQLContextFactory<SpringGraphQLContext>() {
+    override suspend fun generateContextMap(request: ServerRequest): Map<*, Any> = mutableMapOf<Any, Any>()
+        .also { map ->
+            request.headers().firstHeader(FEDERATED_TRACING_HEADER_NAME)?.let { headerValue ->
```
Comment (tapaderster) [inline servers/graphql-kotlin-spring-server/src/main/kotlin/com/expediagroup/graphql/server/spring/execution/SpringGraphQLContextFactory.kt:34]: nit: a comment on why we are adding this header will be great.

**Review by tapaderster — APPROVED** [review]

File: servers/graphql-kotlin-spring-server/src/main/kotlin/com/expediagroup/graphql/server/spring/execution/SpringGraphQLContextFactory.kt (hunk 4 above)
Comment (dariuszkuc) [reply servers/graphql-kotlin-spring-server/src/main/kotlin/com/expediagroup/graphql/server/spring/execution/SpringGraphQLContextFactory.kt:34]: its specified in the KDoc on the class -> `Basic implementation of [SpringGraphQLContextFactory] that populates Apollo tracing header.`

**Review by samuelAndalon — APPROVED** [review]

**Comment by frozenice** [comment]
I'm interested in trying this out. I don't want to build and publish it somewhere for myself and our CI.

@dariuszkuc Do you publish snapshots of the `graphql-kotlin` packages somewhere? Or even better: any idea when this will be released?

**Comment by dariuszkuc** [comment]
@frozenice we just released `6.0.0-alpha.0`

**Pull request merged**